src/search/landmarks/landmark_factory_relaxation.cc

```diff
@@ -15,10 +15,11 @@ void LandmarkFactoryRelaxation::generate_landmarks(const shared_ptr<AbstractTask
     postprocess(task_proxy, exploration);
 }
 
-void LandmarkFactoryRelaxation::postprocess(const TaskProxy &task_proxy, Exploration &exploration) {
+void LandmarkFactoryRelaxation::postprocess(
+    const TaskProxy &task_proxy, Exploration &exploration) {
     lm_graph->set_landmark_ids();
-    mk_acyclic_graph();
     calc_achievers(task_proxy, exploration);
+    mk_acyclic_graph();
```

Comment:
This kind of ordering dependency is a code smell; best not to have it, or at least have the code notice itself when something is wrong via an assertion or similar. The duplicated comments on this are an indication that there is a DRY violation here. Not a problem for now, just a pointer for the future. When I work on code like this, I generally try to leave a comment near the code smell to indicate that we're aware of the problem. (That is, if I have to copy the same comment to multiple files, I'd mention in the comments themselves that this is the case and that we're aware it's a problem.) But no need to change this at this late stage in the issue.

Reply:
We discussed a solution like having a flag `achievers_calculated` or something similar and assert this in `mk_acyclic_graph`. I'm also happy to still add this to this pull request. After all, there was no dependency before we worked on this issue, but now there is one, so we should probably aim to make it less smelly as long as we have ideas to do so.

I don't understand your suggestion concerning the comment(s) yet. Do you mean you would have them point to each other, i.e., the comment for `LandmarkFactoryHM::calc_achievers` should point to the (currently identical) comment for `LandmarkFactoryRelaxation::calc_achievers`? Or to the comment for `mk_acyclic_graph`? Or should it just point out that we are aware that this is not an ideal solution, not pointing to any other code?

Reply:
A flag could be a good idea, but if you prefer to merge with no further changes, that's fine too. After adding a flag, I'd do a few tests locally in debug mode with the relevant configurations to verify that the assertions don't trigger. Regarding the comments: there were two identical comments "Achievers should be calculated..." that can go away once there is an assertion, or be shortened to a single line next to the assertion once we have it. The smell is to have the same comment in multiple places. If that stays, I would change the comments themselves to add something like "DRY: We have the same comment also in XXX; perhaps some refactoring is needed."

Reply:
Depends on the cost/benefit and really requires a case by case decision. I said "could be a good idea" rather than something stronger because I think one can argue either way. The cost in time and space is miniscule and not worth worrying about. The cost in code complexity is hard to judge ahead of time. There could also be a benefit in code complexity because I think we can then remove or drastically shorten several comments. There is also a benefit in preventing mistakes and in explicitly highlighting a design flaw; the very fact that a class does require a certain protocol "first you have to do this, then that" that isn't automatically enforced by things like control flow or type system is generally a sign of bad design. Having this very clear in the code is generally a good thing.